File: solutions/java/secret-handshake/1/src/main/java/HandshakeCalculator.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class HandshakeCalculator {

    List<Signal> calculateHandshake(int number) {
        String bs = Integer.toBinaryString(number);
        int length = bs.length() - 1;

        List<Signal> signals = new ArrayList<>();
        int index = 0;
        for (int i = length; i >= 0; i--) {
            if (index == 5) break;
            if (index == 4) {
                if (bs.charAt(i) == '1') Collections.reverse(signals);
            } else if (bs.charAt(i) == '1') {
                signals.add(Signal.values()[index]);
            }
            index++;
        }

        return signals;
    }

}

File: solutions/kotlin/secret-handshake/1/src/main/kotlin/HandshakeCalculator.kt
object HandshakeCalculator {
    fun calculateHandshake(number: Int): List<Signal> {
        var signals = arrayListOf<Signal>()

        val str = number.toString(2)
        val last = str.length - 1
        var index = 0
        for (i in last downTo 0) {
            if (index == 5) break
            if (index == 4 && str[i] == '1')  {
                return signals.reversed()
            }
            if (str[i] == '1') {
                signals.add(Signal.entries[index])
            }
            index++
        }

        return signals;
    }
}
